File: hbase-common/src/main/java/org/apache/hadoop/hbase/group/GroupAdmin.java
/**
 * Copyright The Apache Software Foundation
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.hadoop.hbase.group;

import org.apache.hadoop.hbase.HostPort;
import org.apache.hadoop.hbase.TableName;
import org.apache.hadoop.hbase.classification.InterfaceAudience;

import java.io.Closeable;
import java.io.IOException;
import java.util.List;
import java.util.Set;

/**
 * Group user API interface used between client and server.
 */
@InterfaceAudience.Private
public interface GroupAdmin extends Closeable {
  /**
   * Gets the group information.
   *
   * @param groupName the group name
   * @return An instance of GroupInfo
   */
  GroupInfo getGroupInfo(String groupName) throws IOException;

  /**
   * Gets the group info of table.
   *
   * @param tableName the table name
   * @return An instance of GroupInfo.
   */
  GroupInfo getGroupInfoOfTable(TableName tableName) throws IOException;

  /**
   * Move a set of serves to another group
   *
   *
   * @param servers set of servers, must be in the form HOST:PORT
   * @param targetGroup the target group
   * @throws java.io.IOException Signals that an I/O exception has occurred.
   */
  void moveServers(Set<HostPort> servers, String targetGroup) throws IOException;

  /**
   * Move tables to a new group.
   * This will unassign all of a table's region so it can be reassigned to the correct group.
   * @param tables list of tables to move
   * @param targetGroup target group
   * @throws java.io.IOException
   */
  void moveTables(Set<TableName> tables, String targetGroup) throws IOException;

  /**
   * Add a new group
   * @param name name of the group
   * @throws java.io.IOException
   */
  void addGroup(String name) throws IOException;

  /**
   * Remove a group
   * @param name name of the group
   * @throws java.io.IOException
   */
  void removeGroup(String name) throws IOException;

  /**
   * Balance the regions in a group
   *
   * @param name the name of the gorup to balance
   * @return
   * @throws java.io.IOException
   */
  boolean balanceGroup(String name) throws IOException;

  /**
   * Lists the existing groups.
   *
   * @return Collection of GroupInfo.
   */
  List<GroupInfo> listGroups() throws IOException;

  /**
   * Retrieve the GroupInfo a server is affiliated to
   * @param hostPort
   * @return
   * @throws java.io.IOException
   */
  GroupInfo getGroupOfServer(HostPort hostPort) throws IOException;
}

File: hbase-shell/src/main/ruby/hbase/group_admin.rb
#
# Copyright The Apache Software Foundation
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

include Java
java_import org.apache.hadoop.hbase.util.Pair

# Wrapper for org.apache.hadoop.hbase.group.GroupAdminClient
# Which is an API to manage region server groups

module Hbase
  class GroupAdmin
    include HBaseConstants

    def initialize(configuration, formatter)
      @admin = org.apache.hadoop.hbase.group.GroupAdminClient.new(configuration)
      @conf = configuration
      @formatter = formatter
    end

    #----------------------------------------------------------------------------------------------
    # Returns a list of groups in hbase
    def listGroups
      @admin.listGroups.map { |g| g.getName }
    end
    #----------------------------------------------------------------------------------------------
    # get a group's information
    def getGroup(group_name)
      group = @admin.getGroupInfo(group_name)
      res = {}
      if block_given?
        yield("Servers:")
      else
        res += v
      end
      group.getServers.each do |v|
        if block_given?
          yield(v.toString)
        else
          res += v.toString
        end
      end
      if block_given?
        yield("Tables:")
      else
        res += v
      end
      group.getTables.each do |v|
        if block_given?
          yield(v.toString)
        else
          res += v.toString
        end
      end
    end
    #----------------------------------------------------------------------------------------------
    # add a group
    def addGroup(group_name)
      @admin.addGroup(group_name)
    end
    #----------------------------------------------------------------------------------------------
    # remove a group
    def removeGroup(group_name)
      @admin.removeGroup(group_name)
    end
    #----------------------------------------------------------------------------------------------
    # balance a group
    def balanceGroup(group_name)
      @admin.balanceGroup(group_name)
    end
    #----------------------------------------------------------------------------------------------
    # move server to a group
    def moveServers(dest, *args)
      servers = java.util.HashSet.new()
      args[0].each do |s|
        servers.add(org.apache.hadoop.hbase.HostPort.valueOf(s))
      end
      @admin.moveServers(servers, dest)
    end
    #----------------------------------------------------------------------------------------------
    # move server to a group
    def moveTables(dest, *args)
      tables = java.util.HashSet.new();
      args[0].each do |s|
        tables.add(org.apache.hadoop.hbase.TableName.valueOf(s))
      end
      @admin.moveTables(tables,dest)
    end
    #----------------------------------------------------------------------------------------------
    # get group of server
    def getGroupOfServer(server)
      @admin.getGroupOfServer(org.apache.hadoop.hbase.HostPort.valueOf(server))
    end
    #----------------------------------------------------------------------------------------------
    # get group of server
    def getGroupOfTable(table)
      @admin.getGroupInfoOfTable(org.apache.hadoop.hbase.TableName.valueOf(table))
    end
    #----------------------------------------------------------------------------------------------
    # get list tables of groups
    def listTablesOfGroup(group_name)
      @admin.listTablesOfGroup(group_name)
    end
  end
end
